src/full/Agda/TypeChecking/Unquote.hs

```diff
@@ -766,3 +775,59 @@ evalTCM v = do
           return x
         _ -> liftTCM $ typeError . GenericDocError =<<
           "Should be a pair: " <+> prettyTCM u
+
+
+------------------------------------------------------------------------
+-- * Trusted executables
+------------------------------------------------------------------------
+
+-- Wen, 2020-06-02
+-- Prototype implementation of `primExec`, which uses `unsafePerformIO`,
+-- since functions defined in `PrimitiveImpl` don't have access to IO.
+
+type ExeArg  = Text
+type StdIn   = Text
+type StdOut  = Text
+type StdErr  = Text
+
+-- | Raise an error if the @--allow-exec@ option was not specified.
+requireAllowExec :: TCM ()
+requireAllowExec = do
+  allowExec <- optAllowExec <$> pragmaOptions
+  unless allowExec $
+    typeError $ GenericError "Missing option --allow-exec"
+
+-- | Convert an @ExitCode@ to an Agda natural number.
+--
+exitCodeToNat :: ExitCode -> Nat
+exitCodeToNat  ExitSuccess    = Nat 0
+exitCodeToNat (ExitFailure n) = Nat (toInteger n)
+
+-- | Call a trusted executable with the given arguments and input.
+--
+--   Returns the exit code, stdout, and stderr.
+--
+tcExec :: ExeName -> [ExeArg] -> StdIn -> TCM Term
+tcExec exe args stdIn = do
+  requireAllowExec
+  exes <- optTrustedExecutables <$> commandLineOptions
+  case Map.lookup exe exes of
+    Nothing -> raiseExeNotFound exe exes
+    Just fp -> do
+      let strArgs    = T.unpack <$> args
+      let strStdIn   = T.unpack stdIn
+      (datExitCode, strStdOut, strStdErr) <- lift $ readProcessWithExitCode fp strArgs strStdIn
+      let natExitCode = exitCodeToNat datExitCode
+      let txtStdOut   = T.pack strStdOut
+      let txtStdErr   = T.pack strStdErr
+      toR <- toTerm
+      return $ toR (natExitCode, (txtStdOut, txtStdErr))
+
+-- | Raise an error if the trusted executable cannot be found.
+--
+--   TODO: this should probably not be a @GenericError@ nor, perhaps, a @TypeError@
+--
+raiseExeNotFound :: ExeName -> Map ExeName FilePath -> TCM a
+raiseExeNotFound exe exes = typeError . GenericError . unlines $
+  ("Could not find trusted executable '" ++ T.unpack exe ++ "' in list:") :
+  [ "  - " ++ T.unpack exe | exe <- Map.keys exes ]
```

Comment:
You could use `GenericDocError` here to get the formatting right. I think `GenericError` does an `fsep . map text . words` on the string.